### PR TITLE
Droth 2285 private road information disappears when geometry changes

### PR DESCRIPTION
**Pull request by FilPinto**

### CODE REVIEW

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/RoadLinkService.scala [hunk 1]
```diff
@@ -1107,6 +1099,95 @@ class RoadLinkService(val vvhClient: VVHClient, val eventbus: DigiroadEventBus,
...
+    def returnEqualAttributes(oldIdsAttributes: Seq[Map[String, String]]): Map[String, String] = {
+      oldIdsAttributes.flatMap { mapToFilter =>
+        mapToFilter.filter { case (attr, value) =>
+          oldIdsAttributes.forall { thisMap =>
+            val option = thisMap.get(attr)
+            if (option.isEmpty) {
+              false
+            } else {
+              option.get == value
+            }
```
Comment (marquesfm) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/RoadLinkService.scala]: We can simplify like this:
```scala
val option = thisMap.get(attr)
option.nonEmpty && option.get == value

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/dao/RoadLinkDAO.scala [hunk 2]
```diff
@@ -318,6 +318,26 @@ object RoadLinkDAO{
       sql"""select name, value from #$table where link_id = $linkId and (valid_to IS NULL OR valid_to > sysdate) """.as[(String, String)].list.toMap
     }
 
+    def getExistingValuesForChanges(linkId: Long, changeTimeStamp: Long) = {
```
[the remark this answers is not in the document]
Comment (FilPinto) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/dao/RoadLinkDAO.scala]: Agree

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/dao/RoadLinkDAO.scala [hunk 3]
```diff
@@ -318,6 +318,26 @@ object RoadLinkDAO{
...
+      sql"""select name, value from #$table where link_id = $linkId and (valid_to IS NULL OR valid_to > sysdate) and adjusted_timestamp < $changeTimeStamp""".as[(String, String)].list.toMap
+    }
+
+    def expireAttributesChange(linkId: Long, username: String, changeTimeStamp: Long): Unit = {
```
[the remark this answers is not in the document]
Comment (FilPinto) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/dao/RoadLinkDAO.scala]: Agree

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/RoadLinkService.scala [hunk 4]
```diff
@@ -1107,6 +1099,95 @@ class RoadLinkService(val vvhClient: VVHClient, val eventbus: DigiroadEventBus,
...
+            attributes.foreach { case (attribute, value) =>
+              LinkAttributesDao.insertAttributeValueChange(change.newId.get, changeUsername, attribute, value, change.vvhTimeStamp)
+            }
+          } else {
+            val roadLinkAttributesRelatedWithThisChange = LinkAttributesDao.getExistingValuesForChanges(change.oldId.get, change.vvhTimeStamp)
+
+            if (roadLinkAttributesRelatedWithThisChange.nonEmpty) {
+              val newIdsOfThisOldId = changesToBeProcessed.filter(thisChange => thisChange.oldId == change.oldId && thisChange.newId.isDefined)
+
+              newIdsOfThisOldId.isEmpty match {
```
[the remark this answers is not in the document]
Comment (FilPinto) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/RoadLinkService.scala]: Agree

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/RoadLinkService.scala (hunk 1 above)
Comment (FilPinto) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/RoadLinkService.scala]: Agree

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/RoadLinkService.scala [hunk 5]
```diff
@@ -1107,6 +1099,95 @@ class RoadLinkService(val vvhClient: VVHClient, val eventbus: DigiroadEventBus,
...
+              val attributes = LinkAttributesDao.getExistingValuesForChanges(thisChange.oldId.get, change.vvhTimeStamp)
+              if (attributes.nonEmpty)
+                LinkAttributesDao.expireAttributesChange(thisChange.oldId.get, changeUsername, thisChange.vvhTimeStamp)
+              attributes
+            }
+
+            val attributes = returnEqualAttributes(oldIdsAttributes)
+            attributes.foreach { case (attribute, value) =>
+              LinkAttributesDao.insertAttributeValueChange(change.newId.get, changeUsername, attribute, value, change.vvhTimeStamp)
+            }
```
[the remark this answers is not in the document]
Comment (FilPinto) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/RoadLinkService.scala]: Agree

**Pull request merged**